subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaLibraryPlugin.java

```diff
@@ -0,0 +1,88 @@
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.compile.JavaCompile;
+
+import static org.gradle.api.plugins.JavaPlugin.USAGE;
+
+/**
+ * <p>A {@link Plugin} which extends the capabilities of the {@link JavaPlugin Java plugin} by cleanly separating
+ * the API and implementation dependencies of a library.</p>
+ */
+public class JavaLibraryPlugin implements Plugin<Project> {
+
+    @Override
+    public void apply(final Project project) {
+        project.getPluginManager().apply(JavaPlugin.class);
+
+        JavaPluginConvention convention = (JavaPluginConvention) project.getConvention().getPlugins().get("java");
+        final ConfigurationContainer configurations = project.getConfigurations();
+        addApiToMainSourceSet(project, convention, configurations);
+    }
+
+    private void addApiToMainSourceSet(final Project project, JavaPluginConvention convention, final ConfigurationContainer configurations) {
+        convention.getSourceSets().getByName("main", new Action<SourceSet>() {
+            @Override
+            public void execute(SourceSet sourceSet) {
+                defineApiConfigurationsForSourceSet(sourceSet, configurations);
+                JavaCompile javaCompile = (JavaCompile) project.getTasks().findByName(sourceSet.getCompileJavaTaskName());
+                project.getArtifacts().add(sourceSet.getApiCompileConfigurationName(), ImmutableMap.of(
+                    "file", javaCompile.getDestinationDir(),
+                    "builtBy", javaCompile));
+            }
+        });
+    }
+
+    private void defineApiConfigurationsForSourceSet(SourceSet sourceSet, ConfigurationContainer configurations) {
+        Configuration apiConfiguration = configurations.maybeCreate(sourceSet.getApiConfigurationName());
+        apiConfiguration.setVisible(false);
+        apiConfiguration.setDescription("API dependencies for " + sourceSet + ".");
+        apiConfiguration.setCanBeResolved(false);
+        apiConfiguration.setCanBeConsumed(false);
+        apiConfiguration.setTransitive(false);
+
+        Configuration apiCompileConfiguration = configurations.maybeCreate(sourceSet.getApiCompileConfigurationName());
+        apiCompileConfiguration.setVisible(false);
+        apiCompileConfiguration.setDescription("API compile classpath for " + sourceSet + ".");
+        apiCompileConfiguration.setCanBeResolved(true);
+        apiCompileConfiguration.setCanBeConsumed(true);
+        apiCompileConfiguration.setTransitive(false);
+        apiCompileConfiguration.attribute(USAGE, "api");
+        apiCompileConfiguration.extendsFrom(apiConfiguration);
+
+        Configuration implementationConfiguration = configurations.maybeCreate(sourceSet.getImplementationConfigurationName());
```

Comment:
The thing is how to effectively deprecate `compile`. If we deprecate it, we need something like the `lock` method introduced here, but with a deprecation warning instead of error. But if we do this, as soon as 3.3 (or 3.4) is out, we're going to pollute all builds out there with this. It's not a small change, it effectively changes almost all the builds around. It's even bigger than deprecating `<<`. Such a big change I'm not sure it's a good idea in the end.

So my idea was more to say "you've been used to `java` and the `compile` configuration, keep going, but now, you have a better solution for libraries, which is `java-library`, and this one has different configurations". We wouldn't need to remove `compile` and instead we make it unusable for that plugin (that what `lock` is for). It would be transparent for all current users and a soft migration path.